src/nupic/algorithms/SpatialPooler.cpp

```diff
@@ -1120,20 +1122,86 @@ void SpatialPooler::updateDutyCyclesHelper_(vector<Real>& dutyCycles,
 
 void SpatialPooler::updateBoostFactors_()
 {
-  for (UInt i = 0; i < numColumns_; i++)
+  if (maxBoost_ > 1.0)
+  {    
+    if (globalInhibition_)
+    {
+      updateBoostFactorsGlobal_();
+    }
+    else
+    {
+      updateBoostFactorsLocal_();
+    }
+  }
+}
+
+void SpatialPooler::updateBoostFactorsGlobal_()
+{
+  Real targetDensity;
+  if (numActiveColumnsPerInhArea_ > 0)
   {
-    if (minActiveDutyCycles_[i] <= 0)
+    UInt inhibitionArea = pow((Real) (2 * inhibitionRadius_ + 1),
+                              (Real) columnDimensions_.size());
+    inhibitionArea = min(inhibitionArea, numColumns_);
+    targetDensity = ((Real) numActiveColumnsPerInhArea_) / inhibitionArea;
+    targetDensity = min(targetDensity, (Real) 0.5);
+  }
+  else
+  {
+    targetDensity = localAreaDensity_;
+  }
+
+  for (UInt i = 0; i < numColumns_; ++i)
+  {
+    Real boostFactor = exp(-(activeDutyCycles_[i] - targetDensity)
+                           * maxBoost_);
+
+    // Avoid floating point mismatches between implementations.
+    boostFactors_[i] = round(boostFactor * 100.0) / 100.0;
+  }
+ 
+}
+
+void SpatialPooler::updateBoostFactorsLocal_()
+{
+  vector<Real> targetDensity(numColumns_, 0);
+  
+  for (UInt i = 0; i < numColumns_; ++i)
+  {
+    UInt numNeighbors = 0;
+    Real localActivityDensity = 0;
+
+    if (wrapAround_)
     {
-      continue;
+      for (UInt neighbor : WrappingNeighborhood(i, inhibitionRadius_,
+                                                columnDimensions_))
+      {
+        localActivityDensity += activeDutyCycles_[neighbor];
+        numNeighbors += 1;
+      }      
     }
-    if (activeDutyCycles_[i] > minActiveDutyCycles_[i])
+    else
     {
-      boostFactors_[i] = 1.0;
-      continue;
+      for (UInt neighbor : Neighborhood(i, inhibitionRadius_,
+                                                columnDimensions_))
+      {
+        localActivityDensity += activeDutyCycles_[neighbor];
+        numNeighbors += 1;
+      }            
     }
-    boostFactors_[i] = ((1 - maxBoost_) / minActiveDutyCycles_[i] *
-                        activeDutyCycles_[i]) + maxBoost_;
+    targetDensity[i] = localActivityDensity / numNeighbors;
```

Comment:
I'm now realizing that this vector doesn't need to exist here either. We could just combine the 2 loops and avoid having to create this vector.

In other words, we could replace this line with:

~~~c++
Real targetDensity = localActivityDensity / numNeighbors;
Real boostFactor = exp(-(activeDutyCycles_[i] - targetDensity) * maxBoost_);

// Avoid floating point mismatches between implementations.
boostFactors_[i] = round(boostFactor * 100.0) / 100.0;
~~~